fix(auto-rules): PeriodicArchiver scans archives on startup

hareetd opened the pull request:
By scanning existing archives on startup, this fix allows pruning of old recordings from previous calls to `PeriodicArchiver` for a given rule. 

Fixes #542 

Depends on #557

hareetd commented:
I requested a review so I know if I'm on the right track before I start updating the corresponding unit tests.

andrewazores commented on src/main/java/io/cryostat/rules/PeriodicArchiver.java:
```diff
@@ -184,4 +200,19 @@ void performArchival() throws InterruptedException, ExecutionException {
     }
+
+    JsonArray getArchivedRecordings() throws InterruptedException, ExecutionException {
+        HttpRequest<Buffer> req = this.webClient.get("api/v1/recordings");
```
We should avoid using HTTP requests to ourselves here - see #543 .

hareetd replied:
Alright, so I should refactor the `RecordingsGetHandler` in such a way that the JSON array can be obtained without using an HTTP request? From a cursory glance at the handler, it uses `FileSystem` to store the data we need, so it looks like I could add and expose a method which extracts the `FileSystem` logic.

andrewazores replied:
Right. The logic should be extracted into something similar to the `RecordingCreationHelper`, for example called the `RecordingArchiveHelper`, which the `RecordingsGetHandler` and the `PeriodicArchiver` can both use for doing things related to adding/listing/deleting files from the archives. The `RecordingArchiveHelper` would take the existing logic from the `RecordingsGetHandler` that creates files in the `FileSystem` and generates their names, and that same logic but in reverse can be used to check the contents of the `FileSystem` and see if anything has a name that matches the format the `RecordingArchiveHelper` uses.

hareetd replied:
Initially, when I saw the changes in @jan-law 's PR, I was thinking I would only have to implement listing archived files in my `RecordingArchiveHelper`, since adding and deleting files in the `PeriodicArchiver` has already been handled by the `RecordingHelper` from that PR.

However, I noticed that the `RecordingHelper` functions are also used by the recording handlers only concerning the target JVM such as `TargetRecordingDeleteHandler`, `TargetRecordingsPostHandler`, etc. 

Would it be worthwhile to keep the logic separate for archive vs. target recordings, i.e. have a `RecordingArchiveHelper` which the `PeriodicArchiver` and the archive related handlers (`TargetRecordingPatchSave`, `RecordingsGetHandler`, and `RecordingDeleteHandler`) all use for adding/listing/deleting files from the archives? 

Then the `RecordingHelper` (potentially renamed to `RecordingTargetHelper`) would no longer be used by the `PeriodicArchiver`, instead by recordings only involving the target?

andrewazores replied:
Sure, that does sound like a distinction probably worth making.

jan-law replied:
@hareetd I can move my related helper functions `saveRecording` ,`deleteRecording` and `writeRecordingToDestination` into a new file `RecordingArchiveHelper`, and also rename `RecordingHelper` to `RecordingTargetHelper`. Then you can move your helper functions into `RecordingArchiveHelper`

hareetd replied:
@jan-law Thank you, that sounds like it'll save me time re-writing the saving/deleting of the archive files. I'm just wondering how it'll work in terms of me getting the changes into my branch in a clean way (i.e. I still want my PR to be clear about its original intention) @andrewazores?

jan-law replied:
@andrewazores since `public Optional<IRecordingDescriptor> getDescriptorByName` is used by both `RecordingTargetHelper` and `RecordingArchiveHelper`, should we make the function static? I think the function output only depends on its arguments.

andrewazores replied:
Rebase your PR on top of Janelle's branch instead of `main` :-) you can then edit your PR's text with "Depends on" and link it to hers that way.

andrewazores replied:
> @andrewazores since `public Optional<IRecordingDescriptor> getDescriptorByName` is used by both `RecordingTargetHelper` and `RecordingArchiveHelper`, should we make the function static? I think the function output only depends on its arguments.

Maybe each of the Helpers should have their own implementation of this function, returning a descriptor (if any) from their respective domains. If some consumer needs to get a descriptor from either domain by name then it can simply use both Helpers and take from them in whichever order of preference makes sense.

andrewazores commented on src/main/java/io/cryostat/rules/PeriodicArchiver.java:
```diff
@@ -97,6 +102,17 @@ public void run() {
+            while (it.hasNext()) {
+                JsonObject entry = (JsonObject) it.next();
+                String fileName = entry.getString("name");
+                if (fileName.contains(this.rule.getRecordingName())) {
```
I think we want a more robust check than `fileName.contains(this.rule.getRecordingName())` - this could lead to a lot of false positive matches if my rule has a really short/common/simple name, or if the rule name happens to equal or be a substring of a target's alias, etc.

andrewazores replied:
There is the `auto_` prefix here that seems to make this less likely, but I think it's still possible we see false positives.

hareetd replied:
I noticed from the HTTP API docs that archived recordings are saved with the following filename format:

[$TARGET_NAME]_[$RECORDING_NAME]_[$DATE]T[$TIME]Z[.$COUNTER][.jfr]

In my opinion, the target name and recording name together can be used to accurately determine a recording as belonging to a specific rule, using a Pattern regex for example.

So now my question is what exactly is TARGET_NAME? Judging from the code used to fulfill a PATCH save request here: 

https://github.com/cryostatio/cryostat/blob/98924a38e879d1284dd4c57369951383f71abf32/src/main/java/io/cryostat/net/web/http/api/v1/TargetRecordingPatchSave.java#L139

it's not simply the `Rule.targetAlias` field so I would have to "calculate" it the same way. This seems redundant however, so maybe I should implement the map as described in the TODO in the link above? That way the map can be used in the `PeriodicArchiver` and when doing a PATCH save. Which class should I store this map in?

andrewazores replied:
That map would help avoid lookups/calculations but it would also need to be repopulated if the Cryostat instance goes down and comes back up later and has pre-existing archives contents, so I don't think it will end up helping you in the end, here.

The targetName part of the saved file is indeed somewhat ambiguous and can't always be mapped exactly back to a target uniquely. We could change this behaviour and always save the file using the service URI of a target instead, which would make them unique/unambiguous.

Another possibility would be to subdivide the archives. Currently all archived recording files are dumped into a flat directory. If we instead created directories within the archive corresponding uniquely to targets/ServiceRefs, then archived files could be placed into those directories and identified as unambiguously belonging to a target as well.

ex. right now we have things like:

```
archive/target1alias_recording1.jfr
archive/target2_recording1.jfr
```

This could be made into:

```
archive/$hashOfTarget1ServiceURI/target1alias_recording1.jfr
archive/$hashOfTarget2ServiceURI/target2_recording.jfr
```

This way we keep the human-friendly filenames but also have a machine-friendly way of grouping and identifying them.

hareetd replied:
Sounds good, I'll take a look into modifying the archive directory structure.

hareetd commented:
@andrewazores After adding Janelle's fork as a remote repo (in order to get her branch), I'm assuming I'd have to wait until she's pushed the required changes before rebasing on top of her branch?

andrewazores commented:
> @andrewazores After adding Janelle's fork as a remote repo (in order to get her branch), I'm assuming I'd have to wait until she's pushed the required changes before rebasing on top of her branch?

You'd have to rebase after each change she makes to her branch anyway, up until eventually her branch is merged into main, at which point you would rebase onto main as well. If her branch doesn't yet contain the changes you need then you can still rebase on top of it for now, and then re-rebase later after her branch is updated with the changes.

hareetd commented:
Before I make more changes, I'm going to rebase on to upstream/main now that Janelle's PR has been merged. During this interactive rebase, I noticed my commit history contains Janelle's commits first (up to a certain point before the branch got merged and deleted) and then my commits (due to the rebasing I did). 

Since upstream//main already contains all of Janelle's commits, should I drop her commits from my branch, keeping only my commits, thereby minimizing merge conflicts during the rebase?

andrewazores commented:
> Before I make more changes, I'm going to rebase on to upstream/main now that Janelle's PR has been merged. During this interactive rebase, I noticed my commit history contains Janelle's commits first (up to a certain point before the branch got merged and deleted) and then my commits (due to the rebasing I did).
> 
> Since upstream/main already contains all of Janelle's commits, should I drop her commits from my branch, keeping only my commits, thereby minimizing merge conflicts during the rebase?

Yes exactly, you should drop all of the commits in your branch that are from the old base. Those changes (roughly) are now in `main`, so your branch should not contain any commits that redo the same changes over again, or else you'll have some strange conflicts to try to resolve where the commits probably just end up empty. git doesn't know what happened to those commits because they got squashed into one larger commit, so when it uses the commit hashes to determine the rebase, it doesn't know that the first series of commits in your branch represent logic changes that are already in the new base.

hareetd commented:
Nice, that makes sense. This a good way for me to more comfortable with Git.

hareetd commented:
Could you take a look at just the `RecordingArchiveHelper` and `PeriodicArchiver` to see if my changes make sense?

andrewazores commented:
> Edit: Also, for the various archived recording handlers (GET, DELETE, etc.) I'm thinking of changing them so that the target service URI is now a request parameter, allowing me to properly index searches in the restructured directory. Does that seem right?

You mean to add the target's service URI as a path parameter on the request? That would be a breaking API change, so it shouldn't be done to the existing V1 handlers.

If I'm understanding correctly, the concern is that ex. `DELETE /api/v1/recordings/:recordingName` doesn't necessarily correlate the request to a target's service URI, since the recording file name may not contain the service URI to be parsed out and used for reference.

This is no different from the current state of the archives really, though - right now we might have `archives/targetA_foo.jfr, archives/targetB_foo.jfr`, but with the subdirectory reorganization here it would be `archives/targetA_hash/targetA_foo.jfr, archives/targetB_hash/targetB_foo.jfr`. So, there's just one additional level of directories to search through to find the matching recording file, unless the client can tell us what `targetA` or `targetA_hash` are. But the problem with that is that the client does not know those values either, since the client can only be presumed to know about `targetA_foo.jfr` because they did a `GET /api/v1/recordings`, and got back the bare list of recording names.

I do think that along with subdividing the archives by target it would make sense to mirror that change in the API, but it would need to be done in V2 handlers. `PATCH 'save' /api/v2/targets/:targetId/recordings/:recordingName` could create an archived recording at `archives/targetId_hash/recordingName_timestamp.jfr`, for example, and then `DELETE /api/v2/targets/:targetId/archives/:filename` could be used for deleting that recording, or `GET` to download it, etc.

If we had that then the unstructured archives used by the V1 handlers could be loosened up further and not require uploads to follow any filename convention, and this could be used to allow users to upload arbitrary JFR files for analysis using jfr-datasource/grafana. They can already do this, really, but the filename convention is a minor barrier that makes that a bit annoying.

Anyway, getting on a tangent. For the scope of this issue/PR, I don't think we want any V1 API behaviour to be changed - everything done in here should be an internal implementation detail. If it sets us up for V2 (or V3) enhancements later like what I outlined above, then great.

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -121,8 +138,50 @@ public void deleteRecording(ConnectionDescriptor connectionDescriptor, String re
                 });
     }
 
+    public List<Map<String, String>> getRecordings() throws Exception {
```
This probably shouldn't return a List of Maps, but instead a List of some more constrained type. The Maps used here always have the same set of keys, so this could/should instead be some class with fields named like those keys, and getters to access them.

Right now I could do `recordingArchiveHelper.getRecordings().get("nonsense key")` and this always returns `null`, but it should just be a compile-time error instead.

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -121,8 +138,50 @@ public void deleteRecording(ConnectionDescriptor connectionDescriptor, String re
+                .map(
+                        file -> {
+                            String serviceUriHash = file.split("/")[0];
+                            String recordingName = file.split("/")[1];
```
Can this be done using the `Path` class rather than `split("/")`? Something like `serviceUriHash = Path.of(file).getParent()`, `recordingName = Path.of(file).getFileName()`.

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -121,8 +138,50 @@ public void deleteRecording(ConnectionDescriptor connectionDescriptor, String re
             JFRConnection connection, IRecordingDescriptor descriptor) throws Exception {
+        URI serviceUri = URIUtil.convert(connection.getJMXURL());
+        Path specificRecordingsPath =
+                recordingsPath.resolve(String.format("%d", serviceUri.hashCode()));
```
`String.format("%d")` seems a little overkill here - `Integer.toString()` might be a better fit.

andrewazores replied:
Or, we could use something other than the direct `int hashCode()` of the serviceUri. It could instead be the base64 encoding of the serviceUri, for example, which is already naturally a String and contains only characters which are legal for filenames/directories, and as an added bonus can be decoded back into a serviceUri (unlike the hashcode, where that information is lost).

hareetd replied:
I'll switch to base64 encoding then, seems to have more pros then the direct hashing, especially the decoding.

andrewazores replied:
https://stackoverflow.com/questions/3945541/base64-encoding-safe-for-filenames

Modified Base64 might be okay, but this discussion makes a good point about using ex. Base32 instead.

andrewazores commented on src/main/java/io/cryostat/rules/PeriodicArchiver.java:
```diff
@@ -146,4 +181,17 @@ public void run() {
+
+    public Future<JsonArray> getArchivedRecordings() throws Exception {
+
+        CompletableFuture<JsonArray> future = new CompletableFuture<>();
```
I'm not sure why this needs to return a `Future` - can't `run()` above just directly call `recordingArchiveHelper.getRecordings()`? Then above the direct return type (currently `List<Map<String, String>>`, but as I said in an earlier comment, should be `List<SomeOtherType>`) can be used, rather than having to go through this JSON encode/decode cycle on top.

hareetd replied:
For the `Future`, I was following the advice here: 

https://github.com/cryostatio/cryostat/pull/557#discussion_r665415792

Should I change it to return a non-Future object? 

The separate method does seem redundant though, especially with the JSON overhead so I'll change that.

andrewazores replied:
Oh, my bad - I think when I wrote that I was thinking that the method belonged to one of the `Helper`s. Since those are part of an internal API, it makes sense for them to be public and to return a `Future` to indicate to callers that these are blocking operations. This `getArchivedRecordings` method and others within the `PeriodicArchiver` are not really part of an internal API, so it's okay for them to be non-`public`, and in that case certainly don't need to return `Future`s or otherwise be marked as long-running operations if they are just internal methods within the class.

andrewazores replied:
So the actual advice I meant to share in that comment would mean that in this case, `RecordingArchiveHelper#getRecordings` should be `public`, and should have a return type of `Future<List<ArchivedRecordingInfo>>` (or whatever that type to replace the `Map<String, String>` gets named).

hareetd replied:
No worries, I'll go ahead and remove the unnecessary `Future` stuff from the `PeriodicArchiver`.

andrewazores commented on src/main/java/io/cryostat/rules/PeriodicArchiver.java:
```diff
@@ -91,6 +97,34 @@ public void run() {
+                Pattern recordingFilenamePattern =
+                        Pattern.compile(
+                                String.format(
+                                        "([A-Za-z\\d-]*)_(%s)_([\\d]*T[\\d]*Z)(\\.[\\d]+)?",
```
Instead of building the regex pattern with the serviceUri inserted into the middle of it, this can be a constant pattern with a capture group where the serviceUri is expected. Then the captured value of that group can be pulled out to compare to the expected serviceUri, or hashed, or whatever else is needed.

andrewazores replied:
It looks like this pattern probably originally came from: https://github.com/cryostatio/cryostat/blob/af47f3f5b08c673cb1df7cae8c739db5b2f597b8/src/main/java/io/cryostat/net/web/http/api/v1/RecordingsPostHandler.java#L76

The matcher and capture group stuff can also be seen in use with that pattern here:

https://github.com/cryostatio/cryostat/blob/af47f3f5b08c673cb1df7cae8c739db5b2f597b8/src/main/java/io/cryostat/net/web/http/api/v1/RecordingsPostHandler.java#L164

https://github.com/cryostatio/cryostat/blob/af47f3f5b08c673cb1df7cae8c739db5b2f597b8/src/main/java/io/cryostat/net/web/http/api/v1/RecordingsPostHandler.java#L169

hareetd replied:
Just to clarify, the regex pattern has the recording name (in this case, `auto_$(Rule.name)`) inserted into the middle, not the serviceUri. The serviceUri hash is compared directly inside the loop, and not with the matcher, since the serviceUri is not part of the overall filename but just the parent directory.

Capturing seems more robust, I'll take a look at those lines, thanks.

andrewazores replied:
Ah right, makes sense. In any case, I think the capture group way is more conventional and so probably easier to read/maintain in the future. As an added bonus it's only one pattern that needs to be compiled at application startup, rather than a new pattern to be compiled on every request, so there should be a (very slight) performance benefit too.

hareetd replied:
Sounds good, I'll implement the capture group regardless.

hareetd commented:
Alright, thanks for the thorough explanation, makes sense.

andrewazores commented on src/main/java/io/cryostat/net/web/http/api/v1/RecordingDeleteHandler.java:
```diff
@@ -102,34 +94,24 @@ public boolean isAsync() {
+                    .message(Map.of("recording", recordingName))
+                    .build()
+                    .send();
+            ctx.response().setStatusCode(200);
```
Just a style thing, but `200` is already the default status code assumed by the `IntermediateResponse` if you don't set one, so explicitly setting it to `200` isn't required but also isn't harmful.

hareetd replied:
I'll remove it then, looks cleaner that way.

andrewazores commented on src/main/java/io/cryostat/net/web/http/api/v1/RecordingDeleteHandler.java:
```diff
@@ -102,34 +94,24 @@ public boolean isAsync() {
-                        });
+        try {
+            recordingArchiveHelper.deleteRecording(recordingName).get();
+            notificationFactory
```
Would it make sense to also extract this notification emission into the archive helper's `deleteRecording` method?

hareetd replied:
Sure, I'll do the same for the patch save.

andrewazores commented on src/main/java/io/cryostat/net/web/http/api/v1/RecordingGetHandler.java:
```diff
@@ -86,28 +85,18 @@ public boolean isAsync() {
-                        });
+        try {
+            Path archivedRecording = recordingArchiveHelper.getRecordingPath(recordingName).get();
+            ctx.response().sendFile(archivedRecording.normalize().toAbsolutePath().toString());
```
Should `getRecordingPath` perhaps already return a normalized, absolute path?

andrewazores commented on src/main/java/io/cryostat/net/web/http/api/v1/RecordingUploadPostHandler.java:
```diff
@@ -140,9 +135,17 @@ public void handleAuthenticated(RoutingContext ctx) throws Exception {
+        } catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+            if (cause instanceof RecordingNotFoundException) {
+                throw new HttpStatusException(404, cause.getMessage(), cause);
```
I think it would be better to pass `e.getMessage()`/`e` into the `HttpStatusException` constructor rather than `cause`, so we aren't losing a layer of context information, right?

andrewazores commented on src/main/java/io/cryostat/net/web/http/api/v1/RecordingsGetHandler.java:
```diff
@@ -111,48 +88,12 @@ public boolean isAsync() {
+            ctx.response().setStatusCode(200);
+            ctx.response().end(gson.toJson(result));
+        } catch (ArchivePathException e) {
+            throw new HttpStatusException(501, e.getMessage());
```
Should `e` be passed to the exception constructor as well to establish the causal link between them?

andrewazores commented on src/main/java/io/cryostat/net/web/http/api/v1/RecordingsPostHandler.java:
```diff
@@ -174,12 +178,18 @@ public void handleAuthenticated(RoutingContext ctx) throws Exception {
                         : Integer.parseInt(m.group(4).substring(1));
 
+        ConnectionDescriptor connectionDescriptor = getConnectionDescriptorFromContext(ctx);
+        Base32 base32 = new Base32();
```
`base32` looks like it could be extracted to a field or constant - doesn't seem like it needs to be instantiated every time.

andrewazores commented on src/main/java/io/cryostat/net/web/http/api/v1/TargetRecordingPatchSave.java:
```diff
@@ -85,8 +86,13 @@ void handle(RoutingContext ctx, ConnectionDescriptor connectionDescriptor) throw
+        } catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+            if (cause instanceof RecordingNotFoundException) {
+                throw new HttpStatusException(404, cause.getMessage(), cause);
```
Same here re. `e` vs `cause`

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -39,90 +39,216 @@
+            future.completeExceptionally(e);
+        }
+
+        return future;
```
This method could be simplified somewhat since the `executeConnectedTask` actually runs synchronously, but the way you've written it lends itself well to the connected tasks possibly being done asynchronously as well, so I'm okay with leaving it as-is.

However, I think it's a bit awkward for consumers of this internal API to receive a `Future` but also have to catch the generic `Exception` when calling the method. I think it would be less surprising if the `catch` block in this method caught all of those `Exceptions, removing it from the method signature, and pushing them into the `future.completeExceptionally` instead. Does that make sense?

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -39,90 +39,216 @@
+        CompletableFuture<Void> future = new CompletableFuture<>();
+
+        try {
+            Path archivedRecording = getRecordingPath(recordingName).get();
```
Maybe the returned `Future` could relay this Path to the caller as well - they may or may not need it, but it's a piece of potentially useful information.

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -39,90 +39,216 @@
-                            this.getDescriptorByName(connection, recordingName);
+        try {
+            Path archivedRecording = null;
+            Boolean recordingFound = false;
```
Can this be `boolean` (lowercase `b`, primitive type) instead?

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -39,90 +39,216 @@
-                });
+                }
+
+                if (recordingFound) {
```
It's easily abused so usually best avoided, but in this case a labelled break:

(see https://docs.oracle.com/javase/tutorial/java/nutsandbolts/branch.html)

could be used instead of this `recordingFound` flag to break out of the outer loop from the inner loop.

andrewazores replied:
Alternatively you could extract this piece of logic out into a small helper method, and use `return` to break the control flow out from the inner loop back up to the callsite.

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -39,90 +39,216 @@
-                        throw new RecordingNotFoundException(recordingName);
+                for (String file : files) {
+                    if (recordingName.equals(file)) {
+                        archivedRecording = recordingsPath.resolve(subdirectory + "/" + file);
```
This part doesn't look right: `recordingsPath.resolve(subdirectory + "/" + file)`

Would this be equivalent to `recordingsPath.resolve(subdirectory).resolve(file)`?

andrewazores commented on src/main/java/io/cryostat/rules/ArchivedRecordingInfo.java:
```diff
@@ -0,0 +1,102 @@
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
+public class ArchivedRecordingInfo {
```
This class looks a lot like the existing `io.cryostat.jmc.serialization.SavedRecordingDescriptor` class, which is apparently now unused. This class probably replaces it, so the `SavedRecordingDescriptor` could be deleted now.

andrewazores commented on src/main/java/io/cryostat/rules/PeriodicArchiver.java:
```diff
@@ -91,59 +97,71 @@ public void run() {
+            // archived or the Cryostat instance was restarted. Since it could be the latter,
+            // populate the array with any previously archived recordings for this rule.
+            if (previousRecordings.isEmpty()) {
+                Base32 base32 = new Base32();
```
Extract from here as well.

andrewazores commented on src/main/java/io/cryostat/rules/PeriodicArchiver.java:
```diff
@@ -91,59 +97,71 @@ public void run() {
+                String serviceUri = serviceRef.getServiceUri().toString();
+                List<ArchivedRecordingInfo> archivedRecordings =
+                        recordingArchiveHelper.getRecordings().get();
+                Iterator<ArchivedRecordingInfo> it = archivedRecordings.iterator();
```
Why is this `Iterator` used instead of a simple for-each loop?

andrewazores commented on src/main/java/io/cryostat/util/ArchivedRecordingInfoSerializer.java:
```diff
@@ -0,0 +1,64 @@
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+
+public class ArchivedRecordingInfoSerializer implements JsonSerializer<ArchivedRecordingInfo> {
```
What does this do that the default Gson serialization for the ArchivedRecordingInfo doesn't? It looks like it drops the `encodedServiceUri` from the serialized representation - if that's the only goal, then simply adding the `transient` keyword to the field declaration in the ArchivedRecordingInfo class would instruct Gson to ignore it in serialization as well, and this class wouldn't be needed.

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -235,25 +235,8 @@
-                }
-            }
+            Path archivedRecording = searchSubdirectories(subdirectories, recordingsPath, recordingName);
             if (!(fs.exists(archivedRecording)
```
Should there be a null-check here too?

andrewazores commented on src/main/java/io/cryostat/net/web/http/api/v1/RecordingGetHandler.java:
```diff
@@ -87,7 +87,7 @@ public void handleAuthenticated(RoutingContext ctx) throws Exception {
         try {
             Path archivedRecording = recordingArchiveHelper.getRecordingPath(recordingName).get();
-            ctx.response().sendFile(archivedRecording.normalize().toAbsolutePath().toString());
+            ctx.response().sendFile(archivedRecording.toString());
```
And here - what if the returned path is null?

hareetd replied:
I'll resolve the null case inside the `getRecordingPath` function.

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -39,90 +39,244 @@
+            for (String file : files) {
+                if (recordingName.equals(file)) {
+                    archivedRecording = parent.resolve(subdirectory).resolve(file).normalize().toAbsolutePath();
+                    return archivedRecording;
```
No need to assign the `archivedRecording` value and then also immediately return it here - you can just directly return the found value.

In fact, you could remove the `archivedRecording` local variable altogether, and simply return the result here. If you get to the end of the method without having returned anything then you can either `return null` or `throw RecordingNotFoundException`.

hareetd commented:
I've implemented all the suggestions above, thank you.

I realized that I had misunderstood the archived report caching process so I added a call to `RecordingArchiveHelper.getRecordingPath(...)` in `ArchivedRecordingReportCache` to fix the issue. However, this resulted in a circular dependency in Dagger because the `RecordingArchiveHelper` calls `ReportService.delete(...)` when deleting an archived recording, where `ReportService` depends on `ArchivedRecordingReportCache`.

In order to resolve this, I extracted the logic for deleting archived recording reports into the `RecordingArchiveHelper`, which allowed me to remove the `ReportService` dependency from the class. However, I was wondering whether it's worth it to make a new class called `ReportArchiveHelper` and put the logic in there?

andrewazores commented on src/main/java/io/cryostat/net/web/http/api/v1/RecordingsPostHandler.java:
```diff
@@ -174,12 +182,17 @@ public void handleAuthenticated(RoutingContext ctx) throws Exception {
                         ? 0
                         : Integer.parseInt(m.group(4).substring(1));
 
+        ConnectionDescriptor connectionDescriptor = getConnectionDescriptorFromContext(ctx);
```
How does this work? `getConnectionDescriptorFromContext` relies on retrieving the `targetId` path parameter, but this API handler doesn't have such a path parameter. I think you'll end up with a `ConnectionDescriptor` with a null targetId and null credentials, no?

hareetd replied:
So the `ConnectionDescriptor` fields are populated using the path parameters, which is why this approach works for `TargetRecordingPatchSave` but not here, right? 

In that case, how would I organize the directory structure since here the recording is being uploaded directly from the client? I believe the only target identifier I have is `TARGET_NAME` in the file format `[$TARGET_NAME]_[$RECORDING_NAME]_[$DATE]T[$TIME]Z[.$COUNTER][.jfr]`.

hareetd replied:
To add on, my changes to the other handlers are all centered around recordings being organized in base 32 encoded `ServiceUri` (i.e. `targetId`) directories.

andrewazores replied:
Right, it works elsewhere in handlers that do have a `targetId` path parameter, but it won't work here unfortunately.

Can you add a special-cased subdirectory for these client re-uploaded recordings? The actual subdirectories would always have names that are base32-encoded ServiceURIs (or maybe base32-encoded host:port pairs), so simply naming the directory an unencoded `"unlabelled"` or something could work, I think. You can upload into that and also merge those into the results when scanning the filesystem for all archived recordings.

At some point in a separate PR we should add a V2 handler that allows for re-uploads of recordings that does have a targetId path parameter so that the uploads can be associated with a particular target. Then we can also make it possible to query for archived recordings by target, and then the web-client can display archived recordings per-target as well.

hareetd replied:
That's a good idea and it's a simple add-on; I'll go ahead and do that. 

As for the V2 handler, I can get started on that if you'd like once I finish this PR and have helped Janelle with the API IT coverage.

andrewazores commented:
> I've implemented all the suggestions above, thank you.
> 
> I realized that I had misunderstood the archived report caching process so I added a call to `RecordingArchiveHelper.getRecordingPath(...)` in `ArchivedRecordingReportCache` to fix the issue.

I'm not sure I follow. What was the issue and how did calling `getRecordingPath` fix it?

> However, this resulted in a circular dependency in Dagger because the `RecordingArchiveHelper` calls `ReportService.delete(...)` after deleting an archived recording, where `ReportService` depends on `ArchivedRecordingReportCache`.
> 
> In order to resolve this, I extracted the logic for deleting archived recording reports into the `RecordingArchiveHelper`, which allows me to remove the `ReportService` dependency from the class. However, I was wondering whether it's worth it to make a new class called `ReportArchiveHelper` and put the logic in there?

That's probably a good idea, but I think it can be left for a later refactor/cleanup after we get the functionality hammered out here.

You also likely could have resolved the circular dependency using `dagger.Lazy`, but extracting common logic to the helpers is also worthwhile to avoid duplication.

hareetd commented:
> I'm not sure I follow. What was the issue and how did calling `getRecordingPath` fix it?
>
The issue was that `ReportIT` was failing. This was because the way archived recording report caching works is that first the path to the archived recording is retrieved and a separate destination path for the report created. Both paths are then passed into the  `SubprocessReportGenerator.exec(...)` method, which dumps a report to the destination path.

Since I had changed the directory structure for archived recordings in the other handlers, report caching was failing since it still used the old way of getting the archived recording path, meaning no archived recording was being found. I fixed this by calling `getRecordingPath`, which uses the new, base 32 encoded `ServiceUri` directory structure to search.  
> 
> That's probably a good idea, but I think it can be left for a later refactor/cleanup after we get the functionality hammered out here.
> 
> You also likely could have resolved the circular dependency using `dagger.Lazy`, but extracting common logic to the helpers is also worthwhile to avoid duplication.

Yeah, the goal was to avoid duplication but good to know `dagger.Lazy` works as well.

andrewazores commented:
Awesome!

```diff
diff --git a/run.sh b/run.sh
index 3cfa1e89..b2c673d6 100755
--- a/run.sh
+++ b/run.sh
@@ -63,7 +63,7 @@ fi
 podman run \
     --pod cryostat \
     --mount type=tmpfs,target=/opt/cryostat.d/conf.d \
-    --mount type=tmpfs,target=/opt/cryostat.d/recordings.d \
+    --mount type=bind,source="$(dirname $0)/recordings",destination=/opt/cryostat.d/recordings.d,relabel=shared,bind-propagation=shared \
     --mount type=tmpfs,target=/opt/cryostat.d/templates.d \
     --mount type=bind,source="$(dirname $0)/truststore",destination=/truststore,relabel=shared,bind-propagation=shared \
     --mount type=bind,source="$(dirname $0)/certs",destination=/certs,relabel=shared,bind-propagation=shared \
```

This minor configuration change lets me easily view the contents of the archive directory.

Then I did:
```
https -f POST :8181/api/v2/rules Content-Type:"multipart/form-data" name=aRule matchExpression="target.alias=='es.andrewazor.demo.Main' && target.annotations.cryostat.PORT==9093" eventSpecifier=template=Continuous archivalPeriodSeconds=10 preservedArchives=3
```

and let it run for a bit, and also manually re-uploaded two recordings into the archives from the web interface.

```
$ tree recordings/
recordings/
├── ONSXE5TJMNSTU2TNPA5HE3LJHIXS6L3KNZSGSL3SNVUTULZPMNZHS33TORQXIORZGA4TGL3KNV4HE3LJ
│   ├── es-andrewazor-demo-Main_auto_aRule_20210806T192754Z.jfr
│   ├── es-andrewazor-demo-Main_auto_aRule_20210806T192804Z.jfr
│   ├── es-andrewazor-demo-Main_auto_aRule_20210806T192814Z.jfr
│   └── es-andrewazor-demo-Main_foo_20210806T192644Z.jfr
└── unlabelled
    ├── es-andrewazor-demo-Main_auto_aRule_20210806T192414Z.jfr
    └── es-andrewazor-demo-Main_ongoing_20210806T191756Z.jfr

2 directories, 6 files
```

Then I stopped and restarted `sh smoketest.sh` and did a `https :8181/api/v1/recordings` and got:

```json
HTTP/1.1 200 OK
content-encoding: gzip
content-length: 239
content-type: text/plain

[
    {
        "downloadUrl": "https://0.0.0.0:8181/api/v1/recordings/es-andrewazor-demo-Main_auto_aRule_20210806T192804Z.jfr",
        "name": "es-andrewazor-demo-Main_auto_aRule_20210806T192804Z.jfr",
        "reportUrl": "https://0.0.0.0:8181/api/v1/reports/es-andrewazor-demo-Main_auto_aRule_20210806T192804Z.jfr"
    },
    {
        "downloadUrl": "https://0.0.0.0:8181/api/v1/recordings/es-andrewazor-demo-Main_foo_20210806T192644Z.jfr",
        "name": "es-andrewazor-demo-Main_foo_20210806T192644Z.jfr",
        "reportUrl": "https://0.0.0.0:8181/api/v1/reports/es-andrewazor-demo-Main_foo_20210806T192644Z.jfr"
    },
    {
        "downloadUrl": "https://0.0.0.0:8181/api/v1/recordings/es-andrewazor-demo-Main_auto_aRule_20210806T192754Z.jfr",
        "name": "es-andrewazor-demo-Main_auto_aRule_20210806T192754Z.jfr",
        "reportUrl": "https://0.0.0.0:8181/api/v1/reports/es-andrewazor-demo-Main_auto_aRule_20210806T192754Z.jfr"
    },
    {
        "downloadUrl": "https://0.0.0.0:8181/api/v1/recordings/es-andrewazor-demo-Main_auto_aRule_20210806T192814Z.jfr",
        "name": "es-andrewazor-demo-Main_auto_aRule_20210806T192814Z.jfr",
        "reportUrl": "https://0.0.0.0:8181/api/v1/reports/es-andrewazor-demo-Main_auto_aRule_20210806T192814Z.jfr"
    },
    {
        "downloadUrl": "https://0.0.0.0:8181/api/v1/recordings/es-andrewazor-demo-Main_auto_aRule_20210806T192414Z.jfr",
        "name": "es-andrewazor-demo-Main_auto_aRule_20210806T192414Z.jfr",
        "reportUrl": "https://0.0.0.0:8181/api/v1/reports/es-andrewazor-demo-Main_auto_aRule_20210806T192414Z.jfr"
    },
    {
        "downloadUrl": "https://0.0.0.0:8181/api/v1/recordings/es-andrewazor-demo-Main_ongoing_20210806T191756Z.jfr",
        "name": "es-andrewazor-demo-Main_ongoing_20210806T191756Z.jfr",
        "reportUrl": "https://0.0.0.0:8181/api/v1/reports/es-andrewazor-demo-Main_ongoing_20210806T191756Z.jfr"
    }
]
```

Thanks for all your hard work on this @hareetd , this looks great.

hareetd commented:
Awesome, it's nice to see the changes visualized in that manner. 

And no problem, it was a great learning experience. Thanks for your detailed reviews/help @andrewazores!

andrewazores commented on src/main/java/io/cryostat/net/reports/ArchivedRecordingReportCache.java:
```diff
@@ -41,111 +41,84 @@
+                                    Duration.ofMillis(TimeoutHandler.TIMEOUT_MS))
+                            .get();
+            f.complete(saveFile);
+        } catch (ExecutionException | InterruptedException e) {
```
I think this can be removed/merged with the `catch (Exception e)` below. Or at least, the `fs.deleteIfExists(dest)` should be performed here as well.

andrewazores commented on src/main/java/io/cryostat/net/web/http/api/v1/RecordingGetHandler.java:
```diff
@@ -86,28 +85,15 @@ public boolean isAsync() {
+        try {
+            Path archivedRecording = recordingArchiveHelper.getRecordingPath(recordingName).get();
+            ctx.response().sendFile(archivedRecording.toString());
+            ctx.response().end();
```
I think `sendFile` already ends the response, so this can be removed.

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -39,90 +39,270 @@
+        } catch (RecordingNotFoundException
+                | IOException
+                | URISyntaxException
+                | FlightRecorderException e) {
```
Can this `catch` be merged with the one below?

hareetd replied:
I was thinking to keep the specific `Exceptions` separate from the generic `Exception` because they'll be wrapped inside an `ExecutionException` but I guess that doesn't matter since the caller can just use `ExceptionUtils.getRootCause(e)` to determine the specific `Exception` thrown.

andrewazores replied:
In both branches the Future is completed exceptionally, so the caller would get an `ExecutionException` on the `.get()` call in either case. So I think collapsing these two clauses together actually won't change anything for callers.

andrewazores commented on src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java:
```diff
@@ -39,90 +39,270 @@
+                    && fs.exists(archivedRecording)
+                    && fs.isRegularFile(archivedRecording)
+                    && fs.isReadable(archivedRecording))) {
+                throw new RecordingNotFoundException("archives", recordingName);
```
`RecordingNotFoundException` will be thrown here if the path exists but is either not `regularFile` (so is a directory, probably) or is unreadable. Should this be a differently named exception in this case, perhaps? Or at least, should the specific reason for the failure (not a file, or not readable) be included in the exception?

hareetd replied:
Good catch, the `ArchivePathException` class I created would be more appropriate here.

andrewazores commented:
I see a test failure with this stack trace in the Cryostat logs, locally:

```
io.vertx.ext.web.handler.impl.HttpStatusException: Internal Server Error
Caused by: java.lang.ClassCastException: class java.io.IOException cannot be cast to class io.cryostat.net.reports.SubprocessReportGenerator$ReportGenerationException (java.io.IOException is in module java.base of loader 'bootstrap'; io.cryostat.net.reports.SubprocessReportGenerator$ReportGenerationException is in unnamed module of loader 'app')
	at io.cryostat.net.web.http.api.v1.TargetReportGetHandler.targetRecordingNotFound(TargetReportGetHandler.java:124)
	at io.cryostat.net.web.http.api.v1.TargetReportGetHandler.handleAuthenticated(TargetReportGetHandler.java:114)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:85)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:62)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

I suspect this is probably the same thing happening and causing the CI run to hang.

hareetd commented:
Thanks, I'll take a look into that. What's weird is that when I run tests locally sometimes they still pass. Also, when they do fail it's not necessarily the exact same tests failing.

andrewazores commented:
Yea, I'm seeing the same thing - some flaky itest results right now.

andrewazores commented:
#634

andrewazores commented:
> Edit: On that note, during an interactive rebase is it fine if I make any changes I see fit when resolving conflicts or should I stick to only pressing the "Accept current changes", "Accept incoming changes", "Accept both changes", etc. commands? I'm worried I might mess something up regarding the commit history if I move stuff around or add/delete too much.

It's best to just resolve the conflicts during the rebase so that each commit in the series can re-apply. If there are any further adjustments required - in this example, updating tests because a method was added to an interface in the branch you're rebasing onto - it's usually easiest/cleanest to do as you did and add that as an additional commit on top. The more you do to adjust the commits in series during conflict resolution, the more likely those changes are to cause further conflicts with the next commits in the series.
